sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java

```diff
@@ -93,13 +121,169 @@ public void populateDisplayData(DisplayData.Builder builder) {
+          try {
+            writer.close();
+            // The writer does not need to be reset, as this DoFn cannot be reused.
+          } catch (Exception closeException) {
+            if (closeException instanceof InterruptedException) {
+              // Do not silently ignore interrupted state.
+              Thread.currentThread().interrupt();
+            }
+            // Do not mask the exception that caused the write to fail.
+            e.addSuppressed(closeException);
+          }
+          throw e;
+        }
+      }
+
+      @Override
+      public void finishBundle(Context c) throws Exception {
+        if (writer != null) {
+          WriteT result = writer.close();
+          c.output(result);
+          // Reset state in case of reuse.
+          writer = null;
+        }
+      }
+
+      @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
+        Write.Bound.this.populateDisplayData(builder);
+      }
+    }
+
+    /**
+     * Like {@link WriteBundles}, but where the elements for each shard have been collected into
+     * a single iterable.
+     *
+     * @see WriteBundles
+     */
+    private class WriteShardedBundles<WriteT> extends DoFn<KV<Integer, Iterable<T>>, WriteT> {
+      private final PCollectionView<WriteOperation<T, WriteT>> writeOperationView;
+
+      WriteShardedBundles(PCollectionView<WriteOperation<T, WriteT>> writeOperationView) {
+        this.writeOperationView = writeOperationView;
+      }
+
+      @Override
+      public void processElement(ProcessContext c) throws Exception {
+        // In a sharded write, single input element represents one shard. We can open and close
+        // the writer in each call to processElement.
+        WriteOperation<T, WriteT> writeOperation = c.sideInput(writeOperationView);
+        LOG.info("Opening writer for write operation {}", writeOperation);
+        Writer<T, WriteT> writer = writeOperation.createWriter(c.getPipelineOptions());
+        writer.open(UUID.randomUUID().toString());
+        LOG.debug("Done opening writer {} for operation {}", writer, writeOperationView);
+
+        try {
+          for (T t : c.element().getValue()) {
+            writer.write(t);
+          }
+        } catch (Exception e) {
+          try {
+            writer.close();
+          } catch (Exception closeException) {
+            if (closeException instanceof InterruptedException) {
+              // Do not silently ignore interrupted state.
+              Thread.currentThread().interrupt();
+            }
+            // Do not mask the exception that caused the write to fail.
+            e.addSuppressed(closeException);
+          }
+          throw e;
+        }
+
+        // Close the writer; if this throws let the error propagate.
+        WriteT result = writer.close();
+        c.output(result);
+      }
+
+      @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
```

Comment:
Agree with comments above. We should keep this for now since the we lose composite transform display data during pipeline translation. Note that this is a general problem with the mechanism for pipeline translation, but Dataflow runner is currently the only running consuming display data.

There are some Dataflow runner tests which verify display data is making it through in primitive transforms as expected. One example: [DataflowCombineTest.java](https://github.com/apache/incubator-beam/blob/master/runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowCombineTest.java#L45)